pkg/bootstrap/render.go

```diff
@@ -19,214 +19,183 @@ import (
-	PriorityClassName         string
-	InstallMode               string
-
-	NodeSelector  map[string]string
-	Tolerations   []corev1.Toleration
-	NodePlacement *operatorv1.NodePlacement
-
-	RegistrationConfiguration *operatorv1.RegistrationConfiguration
-	WorkConfiguration         *operatorv1.WorkAgentConfiguration
-
-	MultipleHubsEnabled              bool
-	DefaultBootstrapKubeConfigSecret BootstrapKubeConfigSecret
-	BootstrapKubeConfigSecrets       []BootstrapKubeConfigSecret
+type KlusterletManifestsConfig struct {
+	chartConfig      *chart.KlusterletChartConfig
+	managedCluster   *clusterv1.ManagedCluster
+	klusterletConfig *klusterletconfigv1alpha1.KlusterletConfig
 }
 
-type ImagePullSecretConfig struct {
-	UseImagePullSecret       bool
-	ImagePullSecretName      string
-	ImagePullSecretData      string
-	ImagePullSecretConfigKey string
-	ImagePullSecretType      corev1.SecretType
+func NewKlusterletManifestsConfig(installMode operatorv1.InstallMode,
+	clusterName string, bootstrapKubeConfig []byte) *KlusterletManifestsConfig {
+	return &KlusterletManifestsConfig{
+		chartConfig: newKlusterletChartConfig(installMode, clusterName, bootstrapKubeConfig),
+	}
 }
 
-type KlusterletManifestsConfig struct {
-	InstallMode operatorv1.InstallMode
-
-	ClusterName                  string
-	KlusterletNamespace          string
-	KlusterletClusterAnnotations map[string]string
-	BootstrapKubeconfig          []byte
-
-	// PriorityClassName is the name of the PriorityClass used by the klusterlet and operator
-	PriorityClassName string
-
-	// Used to determine whether mc is a localcluster.
-	ManagedCluster *clusterv1.ManagedCluster
-
-	klusterletconfig *klusterletconfigv1alpha1.KlusterletConfig
+func newKlusterletChartConfig(installMode operatorv1.InstallMode,
+	clusterName string, bootstrapKubeConfig []byte) *chart.KlusterletChartConfig {
+	allowPrivilegeEscalation := false
+	privileged := false
+	runAsNonRoot := true
+	readOnlyRootFilesystem := true
+
+	chartConfig := &chart.KlusterletChartConfig{
+		ReplicaCount: 1,
+		Images: chart.ImagesConfig{
+			ImagePullPolicy: corev1.PullIfNotPresent,
+			ImageCredentials: chart.ImageCredentials{
+				// set true by default
+				CreateImageCredentials: true,
+			},
+		},
+		// the release ns and the agent ns are same, so disable to generate ns.
+		CreateNamespace: true,
+		PodSecurityContext: corev1.PodSecurityContext{
+			RunAsNonRoot: &runAsNonRoot,
+		},
+		SecurityContext: corev1.SecurityContext{
+			Capabilities: &corev1.Capabilities{
+				Drop: []corev1.Capability{"ALL"},
+			},
+			Privileged: &privileged,
 
-	generateImagePullSecret bool // by default is true, in hosted mode, it will be set false
-}
+			RunAsNonRoot:             &runAsNonRoot,
+			ReadOnlyRootFilesystem:   &readOnlyRootFilesystem,
+			AllowPrivilegeEscalation: &allowPrivilegeEscalation,
+		},
+		Resources: corev1.ResourceRequirements{
+			Limits: corev1.ResourceList{
+				corev1.ResourceMemory: resource.MustParse("2Gi"),
```

Comment:
in the last release, @haoqing0110 removed the limits for all components, so I am not sure if this should be removed.

Reply:
yes, we removed the limit for the hub components, which is a best practice for scaling. Is this line for klusterlet? it's fine to leave it for the spoke I think.

Reply:
yes, this is for spoke cluster.